### ISSUE TITLE
analizeCityDistances is "tail recursive"

### ISSUE THREAD

**Issue by cohadar** [justradojko/CityDistances#5]
Great idea, and I realize you did it to avoid having 2 for loops, but it is not what I had in mind.
Since tail recursion is inside a for loop the algorithm is still O(n^2)
How can you use only one for loop (and no recursion) to find distances?
Hint: use Sort before for.

**Comment by justradojko** [comment]
Latest code has O(n^2) as well, or in the worst case scenario n(n+1)/2.

Is it possible to achieve anything lower?

**Comment by cohadar** [comment]
Yes, O(n) is possible.

**Comment by cohadar** [comment]
Correction, best time complexity is O(n log n)

**Comment by justradojko** [comment]
Thing that comes to my mind is to loop through cityList (which is complexity O(n)). Then, for each element I would create 6h window, where end point of the window would be found using binary search (complexity O(log n)). But the extra thing is that first element needs to be paired with every other city inside this window which will create complexity bigger then O(n log n).

In my example of 1000 cities, output should have around 200 000 city pairs.

**Comment by cohadar** [comment]
What is the time complexity of sort?

**Comment by justradojko** [comment]
It's n log n. I still can't figure out how to solve this with just one for loop. Maybe I'm not that good in algorithms as you think :/

**Comment by cohadar** [comment]
Sort a CityRecord by time, this will ensure that cities with times
less than 6 will be immediately before the current city of iteration.

That way when you for over a CityRecord list you only need to check a
few "close" array locations.

**Comment by justradojko** [comment]
Last commit is doing exactly that. Snap from analizeCityDistances() method: 

``` java
for (int i = 0; i < cityList.size() - 1; i++) {
            int j = i + 1;

            while ( (t = Math.abs( cityList.get(i).getTimeInMinutes() - cityList.get(j).getTimeInMinutes() )) < (MAX_HOURS * 60)) {
                closeCities.add(new CloseCities(convertMinToTime(t), cityList.get(i).getCityName(), cityList.get(j).getCityName()));
                if (j < cityList.size() - 1)
                    j++;
                else
                    break;
            }
        }
```

In the worst case scenario, where all cities from the list are within 6h distance, algorithm would iterate n(n+1)/2 times which is O(n^2).

**Comment by cohadar** [comment]
Ok, didn't see that you had a sort in loadData function.
Do not put functionality in a function just because it goes
immediately before or after code in that function. Or rename it to
loadDataAndSort...or better just extract sort to go function.

Anyways my fault is for not reading code carefully, yours is for not
recognizing you had good solution already, read this:
https://en.wikipedia.org/wiki/Average-case_complexity

So basically when you sort array you have:
Average: O(n log n), Worst: O(n^2)
but when you do not sort you have:
Average: O(n^2), Worst: O(n^2)

Also note that when you have O(n^2) in sorted solution it is because
data is such, not because algorithm is bad, O(n^2) is in fact optimal
solution for a graph of cities all in 6h range because you want to
print all pairs anyways.

**Comment by justradojko** [comment]
Thanks for inputs.

Just a little note. Java documentation says that Collection sort uses merge sort algorithm which has complexity in worst case O(n log n).

Is there some technique for determining best possible complexity for a given problem or it's defined by problem type and experience? 

EDIT: I'll google that in order not waste your time more then needed :)

**Comment by cohadar** [comment]
Just to be clear: average and worst time complexities I mentioned above are for full "city distances < 6" algorithm, not for sorting.

Algorithm has 2 parts: sort and for loop to find < 6
Merge sort has complexities: Best: O(n), Average: O(n log n), Worst: O(n log n)
For loop in < 6 has complexities: Best: O(n), Average: O(n), Worst: O(n^2)
So this gives:
Best: O(n) + O(n) = O(n)
Average: O(n log n) + O(n) = O(n log n)
Worst: O(n log n) + O(n^2) = O(n^2)

In theoretically best case(almost never happens here): your input is an already sorted list of cities with time distances that are increments of 6.
In average case, it is an unsorted list of cities with uniform random distribution of times.
In worst case it is an unsorted list of cities with all times clustering in small time range < 6.

When you learn complexities of standard algorithm building blocks, like sorting, binary-search, etc. it will become easier to simply see how your algorithm is constructed from parts and how that parts affect complexity.
Practice makes perfect.
